Add optional sort_order property to actions

zentron opened the pull request:
A behavioural "bug" exists due to the way in which different actions are exposed internally before submitting over the wire. Before we update the plugin framework and leverage some new schema capabilities this change provides a temporary work around.

Actions now provide a `sort_order` property that allows for ordering the different actions within a step. This value is optional and can be omitted entirely, but it is recommended that if setting this property for any, then it should be set for all.

Fixes https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/issues/445

zentron commented on octopusdeploy/schema_deployment_step.go:
```diff
@@ -42,60 +44,41 @@ func expandDeploymentStep(flattenedStep map[string]interface{}) *deployments.Dep
 		step.Properties["Octopus.Action.MaxParallelism"] = core.NewPropertyValue(windowSize.(string), false)
 	}
 
-	if v, ok := flattenedStep["action"]; ok {
```
There's a whole bunch of repetitive code here. Have pulled it out into local functions

zentron commented on octopusdeploy/schema_deployment_step.go:
```diff
@@ -42,60 +44,41 @@ func expandDeploymentStep(flattenedStep map[string]interface{}) *deployments.Dep
-			action := expandRunScriptAction(tfAction.(map[string]interface{}))
-			step.Actions = append(step.Actions, action)
+				// Pull out the sort_order if it exists. This is used to sort the actions later
+				if posn, ok := flattenedAction["sort_order"].(int); ok && posn >= 0 {
```
We we read the manifest resources, we create a local lookup that we later use to reorganize the actual Octopus resources. Since name is required and must be unique it serves as a reasonable key

zentron commented on octopusdeploy/schema_deployment_step.go:
```diff
@@ -108,45 +91,59 @@ func flattenDeploymentSteps(deploymentSteps []*deployments.DeploymentStep) []map
 		for i := range deploymentStep.Actions {
 			switch deploymentStep.Actions[i].ActionType {
 			case "Octopus.KubernetesDeploySecret":
-				flattenedDeploymentSteps[key]["deploy_kubernetes_secret_action"] = []interface{}{flattenDeployKubernetesSecretAction(deploymentStep.Actions[i])}
```
I think there was\is actually a bug here too, in that it would only ever store the last action per action type rather than the full array. Thats why the `terraform plan` when multiple actions were involved would always say changes are present, even if none were

johnsimons replied:
yes, I have come across that one too, nice

johnsimons commented on octopusdeploy/schema_deployment_action.go:
```diff
@@ -244,6 +244,12 @@ func getActionSchema() (*schema.Schema, *schema.Resource) {
 				ValidateDiagFunc: warnIfIncludesRunOnServer(),
 			},
+			"sort_order": {
+				Description: "Order used by terraform to ensure correct ordering of actions. This property must be either omitted from all actions, or provided on all actions",
```
Given the order is random, would it make more sense to enforce this if `actions > 1` ?

johnsimons replied:
yes, breaking change, but IMHO, it is way safer, and hence worth breaking it

zentron replied:
🤔 part of why i left it is TF has really bad capacity for custom validation. If you run `terraform validate` it seems like you can only add custom validation for single property values, not for complex logic based on multiple properties. 
Ill take another look and see how useful the failure is if it can at least be caught during `plan`

johnsimons replied:
you want the [new framework ](https://developer.hashicorp.com/terraform/plugin/framework/resources/validate-configuration)

zentron replied:
Yep once we are using that we will potentially have other options available.

johnsimons commented on octopusdeploy/schema_deployment_step.go:
```diff
@@ -42,60 +44,41 @@ func expandDeploymentStep(flattenedStep map[string]interface{}) *deployments.Dep
-		for _, tfAction := range v.([]interface{}) {
-			action := expandDeployKubernetesSecretAction(tfAction.(map[string]interface{}))
-			step.Actions = append(step.Actions, action)
+	// Now that we have extracted all the steps off each of the properties into a single array, sort the array by the sort_order if provided
```
so this does not seem to validate that the user hasn't made a mistake a typed the same order number twice?

zentron replied:
Yeah I had considered doing some validation around that. Rather than failing its easier to just log it with a warning\message. 
The current terraform provider library is actually really bad at allowing complex validation prior to apply. Im hoping the new one fixes some of these things up.

zentron replied:
Ill at least put a log around it. Its a shame there isnt simple linq-style syntax in go :)

johnsimons commented on octopusdeploy/schema_deployment_step.go:
```diff
@@ -1,6 +1,8 @@
 package octopusdeploy
 
 import (
+	"log"
```
I wonder if `github.com/hashicorp/terraform-plugin-log/tflog` is a better package to use

zentron replied:
I did use that at first but it required passing around the context object around. From what I have read there isnt really much difference between them. At the very least we seem to use both right now so im not sure which is "the best" TBH

johnsimons commented on octopusdeploy/schema_deployment_step.go:
```diff
@@ -42,60 +44,41 @@ func expandDeploymentStep(flattenedStep map[string]interface{}) *deployments.Dep
-			step.Actions = append(step.Actions, action)
+				// Pull out the sort_order if it exists. This is used to sort the actions later
+				if posn, ok := flattenedAction["sort_order"].(int); ok && posn >= 0 {
+					name := flattenedAction["name"].(string)
```
we already declared `name`, can we just reuse it?

johnsimons replied:
my bad, never mind

johnsimons commented on octopusdeploy/schema_deployment_action.go:
```diff
@@ -244,6 +244,12 @@ func getActionSchema() (*schema.Schema, *schema.Resource) {
 				Type:             schema.TypeMap,
 				ValidateDiagFunc: warnIfIncludesRunOnServer(),
 			},
+			"sort_order": {
```
Also, I understand zero is ok for devs, but I wonder if it would be nice for this to start at `1`?

zentron replied:
Honestly as long as its sequential I dont really care. If someone puts 0 there's currently no way to differentiate that with not provided, since integer it defaults to 0.

johnsimons replied:
you can add a validator to this field

zentron replied:
We support 0 on other fields so for the sake of consistency im going to leave it as-is for now

johnsimons reviewed (changes requested):
Nice refactoring @zentron 
I added a few comments along the way.

johnsimons commented:
Could we write some unit tests for the sorting @zentron ?

zentron commented:
> Could we write some unit tests for the sorting @zentron ?

👍 Yeah @johnsimons, I plan on adding them in once we are happy with the approach

zentron commented:
Added some additional tests to validate the `sort_order` "fixes" the problem

### [Before](https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/actions/runs/5384045248/jobs/9771507555?pr=529#step:14:381) 
Added [additional test assertions](https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/pull/525/commits/bc8871efd59c50cd534f51f292520a611d485f1f) to validate before and after.
![image](https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/assets/1830666/8c3582e5-3629-4339-9df7-f570b08aed36)

### [After](https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/actions/runs/5384039034/jobs/9771495539?pr=525#step:14:25)
![image](https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/assets/1830666/044a9044-9c4d-4daf-abbd-ff921571feaf)

johnsimons reviewed (approved):
@zentron I tested this from a user's perspective.

I did the following.
I had an existing project with the following deployment process:
```hcl
resource "octopusdeploy_deployment_process" "deployment_process_project_noopterraform" {
  project_id = octopusdeploy_project.deploy_frontend_project.id

  step {
    condition           = "Success"
    name                = "Hello world (using PowerShell)"
    package_requirement = "LetOctopusDecide"
    start_trigger       = "StartAfterPrevious"
    properties          = {}
    target_roles        = ["bread"]

    run_script_action {
      name          = "Pre Script Action"
      features      = []
      run_on_server = false
      script_syntax = "Bash"
      script_body   = <<-EOT
          echo "Pre Script Action"
        EOT
    }

    action {
      action_type                        = "Octopus.Script"
      name                               = "Hello world (using PowerShell)"
      condition                          = "Success"
      run_on_server                      = true
      is_disabled                        = false
      can_be_used_for_project_versioning = false
      is_required                        = true
      worker_pool_id                     = ""
      properties = {
        "Octopus.Action.Script.ScriptSource" = "Inline"
        "Octopus.Action.Script.ScriptBody"   = "Write-Host 'Hello world, using PowerShell'\n\n#TODO: Experiment with steps of your own :)\n\nWrite-Host '[Learn more about the types of steps available in Octopus](https://oc.to/OnboardingAddStepsLearnMore)'"
        "Octopus.Action.Script.Syntax"       = "PowerShell"
      }
      environments          = []
      excluded_environments = []
      channels              = []
      tenant_tags           = []
      features              = []
    }

    run_script_action {
      name          = "Post Script Action"
      features      = []
      run_on_server = false
      script_syntax = "Bash"
      script_body   = <<-EOT
          echo "Post Script Action"
        EOT
    }
  }
}
```
I then updated the provider to this version, but looking at the output, it still contains the bug.
![image](https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/assets/122651/db3a8ca2-0730-4ee6-a4eb-944bb39bd903)

The CLI did not warn me about it either:
![image](https://github.com/OctopusDeployLabs/terraform-provider-octopusdeploy/assets/122651/2e289c3b-f37a-45d9-93fa-0158654ae929)

I know you intend to make this backwards compatible and hence make this `sort_order` optional, but IMO, we are, in a way, not making customers fall into the pit of success.
Are you sure you want to keep this backwards compatible, knowing it can cause this behaviour? To me, this is a dangerous bug with unknown consequences that is worth breaking compatibility.

Anyway, I thought I give you my last plea to reconsider this fix 😄 

I reviewed your changes, and happy with the tests 👍
